internal/bootstrap/bootstrap_provider.go

```diff
@@ -39,7 +40,12 @@ type GitProviderBootstrapper struct {
 
 	owner      string
 	repository string
-	personal   bool
+
+	// stash needs an org key and a repo slug
+	ownerKey       string
+	repositorySlug string
```

Comment:
Introducing implementation specific workarounds in a solution that was written to deal with things in a generic manager makes me feel a bit uncomfortable. In addition, this struct was designed to be configured through options, but is now used to temporary store data gathered during the bootstrap itself.

An improvement within the context of this PR would be to at least move the values to variables (or a struct) that is moved separately through the (private) methods that require the data. However, based on things like:

```go
	var identity, repository string
	if b.ownerKey != "" {
		identity = b.ownerKey
	} else {
		identity = fmt.Sprintf("~%s", rs.GetIdentity())
	}
```

I am wondering if the core of the issue isn't the result of "design awkwardness" on the side of the go-git-providers code, due to e.g. not slightly changing an interface because of the impact it would have on underlying logic from implementations. As from the above, it looks like an "identity" can now be two different things, and the consumer of the package needs to deal with this double, instead of the lib solving it for them.

If this happens to be true, please shout (with some specific pointers in the go-git-providers code that resulted in this), and I'll try to be of helping settling it on the lib side.

Reply:
Thanks for reviewing this 🙏 

I agree with your concerns. Let me try to explain how it came to this.

In go-git-provider the interfaces `IdentityRef` and `RepositoryRef` defines how to handle org and repos. There are  implementations for repos and orgs with methods to get an `org` or `repo` name. 

Then there are the client interface that tell you exactly what a specific method should get and return.
https://github.com/fluxcd/go-git-providers/blob/bbf0f49cbacbc7f17ad405dd79746f966d60ddc1/gitprovider/client.go#L85

The interfaces are designed for github. Stash use both org name and org key for example in its api. It would need a redesign to take that into account. So I decide the best I could do without redefining the interfaces, was to complement the interfaces with `Keyer` and `Slugger`
https://github.com/fluxcd/go-git-providers/blob/bbf0f49cbacbc7f17ad405dd79746f966d60ddc1/gitprovider/repositoryref.go#L63

https://github.com/fluxcd/go-git-providers/blob/bbf0f49cbacbc7f17ad405dd79746f966d60ddc1/gitprovider/repositoryref.go#L81

Then i implemented the methods for `OrganizationRef` `OrgRepositoryRef` and `UserRepositoryRef`.

Reply:
I have updated the code with the last `go-git-provider` pr. Also I have added a new `struct` keep the repo once created, and reuse it for reconciliation.